docs/proposals/petset.md

```diff
@@ -0,0 +1,533 @@
+
+
+```
+type PetSetSpec struct {
+    // Number of instances in the set
+    Replicas *int
+
+    // The template to use for the pod in the set
+    Template *PodTemplateSpec
+}
+
+// Proposal 1a: Use the name of the pod as the identity, use downward API for getting identity
+// Downside: all fields that want to leverage the identity need equivalent valueFrom semantics
+{
+    "spec": {
+        "template": {
+            "metadata": {"generateName": "pod-name"}, // assume identity is automatically added to pod name
+            "spec": {
+                "containers": [{
+                    "env": [{
+                        "name": "IDENTITY",
+                        "valueFrom": {"fieldRef": {"fieldPath": "metadata.name"}
+                        }
+                    }]
+                }]
+            }
+        }
+    }
+}
+
+// Proposal 1b: Add an annotation to each pod when the controller creates it
+// Downside: all fields that want to leverage the identity need equivalent valueFrom semantics
+{
+    "spec": {
+        "template": {
+            "spec": {
+                "containers": [{
+                    "env": [{
+                        "name": "IDENTITY",
+                        "valueFrom": {"fieldRef": {"fieldPath": "metadata.annotations[alpha.kubernetes.io/identity]"}}
+                    }]
+                }]
+            }
+        }
+    }
+}
+
+// Proposal 1c: Add a new field to pod spec that is write-once/read-many - identity
+// Downside: all fields that want to leverage the identity need equivalent valueFrom semantics
+{
+    "spec": {
+        "template": {
+            "spec": {
+                "identity": "", // set on creation
+                "containers": [{
+                    "env": [{
+                        "name": "IDENTITY",
+                        "valueFrom": {"fieldRef": {"fieldPath": "spec.identity"}}
+                    }]
+                }]
+            }
+        }
+    }
+}
+
+// Proposal 1d: support template transformation of a field path
+// Downside: possibly too flexible, can conflict with existing env resolution semantics, bypasses downward API
+{
+    "spec": {
+        "template": {
+            "metadata": {"name": "pod-name-$(IDENTITY)"},
+            "spec": {
+                "containers": [{
+                    "env": [{
+                        "name": "IDENTITY",
+                        "value": "$(IDENTITY)"
+                    }]
+                }]
+            }
+        }
+    }
+}
+
+// Proposal 1e: specify merge-patch semantics with some variable replacement
+// Downside: merge patch is inflexible to structural template edits
+{
+    "spec": {
+        "template": {
+            "metadata": {},
+            "spec": {
+                "containers": [{
+                    "env": [{
+                        "name": "IDENTITY",
+                    }]
+                }]
+            }
+        },
+        "templateTransform": [
```

Comment:
I agree this is more complex, but it is the most future-safe for things like non-string types and expressions.  I don't know if the patch syntax has value or not.